File: app/src/main/java/com/nawin/androidmvparchitecture/utils/NonNullMapper.java
package com.nawin.androidmvparchitecture.utils;

import com.nawin.androidmvparchitecture.data.error.FailedResponseException;
import com.nawin.androidmvparchitecture.data.model.api.BaseResponse;

import io.reactivex.Single;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Function;

public class NonNullMapper<T> implements Function<BaseResponse<T>, Single<T>> {
    @Override
    public Single<T> apply(@NonNull BaseResponse<T> baseResponse) throws Exception {
        if (baseResponse.getStatusCode() != 1)
            return Single.error(new FailedResponseException(baseResponse.getStatusCode(), "message"));
        T item = baseResponse.getResponse();
        if (item == null)
            return Single.error(new NullPointerException("BaseResponse.Response == null"));
        else
            return Single.just(item);
    }
}

File: app/src/main/java/com/nawin/androidmvparchitecture/utils/NonNullMapper.kt
package com.nawin.androidmvparchitecture.utils

import com.nawin.androidmvparchitecture.data.error.FailedResponseException
import com.nawin.androidmvparchitecture.data.model.api.BaseResponse
import io.reactivex.Single
import io.reactivex.annotations.NonNull
import io.reactivex.functions.Function

/**
 * Created on 11/29/17.
 */
class NonNullMapper<T> : Function<BaseResponse<T>, Single<T>> {
    @Throws(Exception::class)
    override fun apply(@NonNull baseResponse: BaseResponse<T>): Single<T> {
        if (baseResponse.statusCode != 1)
            return Single.error(FailedResponseException(baseResponse.statusCode, baseResponse.statusMessage))
        val item = baseResponse.response
        return if (item == null)
            Single.error(NullPointerException("BaseResponse.Response == null"))
        else
            Single.just(item)
    }
}